go/internal/feast/onlinestore/cassandraonlinestore.go

```diff
@@ -0,0 +1,299 @@
+	if keyspaceStr, ok = keyspace.(string); !ok {
+		return nil, fmt.Errorf("failed to convert keyspace to string: %+v", keyspaceStr)
+	}
+
+	protocolVersion, ok := onlineStoreConfig["protocol_version"]
+	if !ok {
+		protocolVersion = 4.0
+		log.Warn().Msg("protocol_version not specified: Using 4 instead")
+	}
+	protocolVersionInt := int(protocolVersion.(float64))
+
+	redisTraceServiceName := os.Getenv("DD_SERVICE") + "-cassandra"
+	if redisTraceServiceName == "" {
+		redisTraceServiceName = "cassandra.client" // default service name if DD_SERVICE is not set
+	}
+	store.clusterConfigs = gocqltrace.NewCluster(cassandraHostsStr, gocqltrace.WithServiceName(redisTraceServiceName))
+	// TODO: Figure out if we need to offer users the ability to tune the timeouts
+	//store.clusterConfigs.ConnectTimeout = 1
+	//store.clusterConfigs.Timeout = 1
+	store.clusterConfigs.ProtoVersion = protocolVersionInt
+	//store.clusterConfigs.Consistency = gocql.Quorum
+	store.clusterConfigs.Keyspace = keyspaceStr
+	loadBalancingPolicy, ok := onlineStoreConfig["load_balancing"]
+	if !ok {
+		loadBalancingPolicy = gocql.RoundRobinHostPolicy()
+		log.Warn().Msg("No load balancing policy selected; setting Round Robin Host Policy")
+	}
+	loadBalancingPolicyStr, _ := loadBalancingPolicy.(string)
+	if loadBalancingPolicyStr == "DCAwareRoundRobinPolicy" {
+		store.clusterConfigs.PoolConfig.HostSelectionPolicy = gocql.RoundRobinHostPolicy()
+	} else if loadBalancingPolicyStr == "TokenAwarePolicy(DCAwareRoundRobinPolicy)" {
+		// Configure fallback policy if unable to reach the shard
+		fallback := gocql.RoundRobinHostPolicy()
+		// If using ScyllaDB and setting this policy, this makes the driver shard aware to improve performance
+		store.clusterConfigs.PoolConfig.HostSelectionPolicy = gocql.TokenAwareHostPolicy(fallback)
+		if config.OnlineStore["type"] == "scylladb" {
+			store.clusterConfigs.Port = 19042
+		} else {
+			store.clusterConfigs.Port = 9042
+		}
+	}
+
+	store.clusterConfigs.Authenticator = gocql.PasswordAuthenticator{Username: usernameStr, Password: passwordStr}
+	createdSession, err := store.clusterConfigs.CreateSession()
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to the ScyllaDB database")
+	}
+	store.session = createdSession
+	return &store, nil
+}
+
+func (c *CassandraOnlineStore) getFqTableName(tableName string) string {
+	return fmt.Sprintf(`"%s"."%s_%s"`, c.keyspace, c.project, tableName)
+}
+
+func (c *CassandraOnlineStore) getCQLStatement(tableName string, featureNames []string, nKeys int) string {
+	// TODO: Compare with multiple single-key concurrent queries like in the Python feature server
+	keyPlaceholders := make([]string, nKeys)
+	for i := 0; i < nKeys; i++ {
+		keyPlaceholders[i] = "?"
+	}
+
+	// this prevents fetching unnecessary features
+	quotedFeatureNames := make([]string, len(featureNames))
+	for i, featureName := range featureNames {
+		quotedFeatureNames[i] = fmt.Sprintf(`'%s'`, featureName)
+	}
+
+	return fmt.Sprintf(
```

Comment:
IN has limit on 100 keys. We should see how we can handle keys with more than 100 features.

Reply:
Yeah, the way to go once we reach that point is to partition requests and execute them in parallel, adds complexity but would work.